### PR TITLE
implemented enable/disable SEND option while an operation is in progress

### LINKED ISSUE #4332: Nothing happens when offline and tapping the SEND button in user Invite screen
### Expected behavior
In the user Invite Screen, tapping "SEND" while typing a username should bring up some form of error message to denote the inability to perform the action while offline

### Actual behavior
Tapping SEND in those conditions only buttonizes the username and then nothing happens. 

### Steps to reproduce the behavior

* Go into the user Invites screen by visiting MySite > People and tapping the "+" button on the actionbar
* Put the device to offline mode
* Start typing a username and before hitting "enter"
* Tap the "SEND" action
* The username mutates into a "button" like widget
* Nothing else happens.

##### Tested on [device], Android [version]

Nexus 5X emulator running Android 6.0

### PR DESCRIPTION
**Pull request by mzorz**
Fixes #4332

To test: 
- Go into the user Invites screen by visiting MySite > People and tapping the "+" button on the actionbar
- Start typing a username and before hitting "enter"
- Put the device to offline mode
- Tap the "SEND" action

The SEND button is disabled right after you tap on it and validations are passed, and is re-enabled once the operation is finished (either successfully or unsuccessfully).

This comes particularly handy in the case of not having good network coverage (not necessarily when no network at all, but rather an intermittent connection or delayed response from the server), as it may take several seconds for the action to complete, or even fail by timeout.

Needs review: @hypest

### CODE REVIEW

File: WordPress/src/main/java/org/wordpress/android/ui/people/PeopleInviteFragment.java [hunk 1]
```diff
@@ -57,6 +58,7 @@
     private Map<String, View> mUsernameErrorViews = new Hashtable<>();
     private String mRole;
     private String mCustomMessage = "";
+    private boolean mInviteOperationInProgress = false;
```
Comment (hypest) [inline WordPress/src/main/java/org/wordpress/android/ui/people/PeopleInviteFragment.java]: This seems like an addition to the Fragment's state so, I suspect it needs to be retained on rotation (via `onSaveInstanceState()`). Ignore this if saving is unneeded. Thanks!

File: WordPress/src/main/java/org/wordpress/android/ui/people/PeopleInviteFragment.java [hunk 2]
```diff
@@ -423,6 +431,10 @@ private void checkAndSend() {
             return;
         }
 
+        if (!NetworkUtils.checkConnection(getActivity())) {
+            return;
```
Comment (hypest) [inline WordPress/src/main/java/org/wordpress/android/ui/people/PeopleInviteFragment.java]: I'd suggest we let the user know that the action was not performed due to lack of connectivity or something. Wdyt?

File: WordPress/src/main/java/org/wordpress/android/ui/people/PeopleInviteFragment.java [hunk 3]
```diff
@@ -74,6 +76,12 @@ public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
         super.onCreateOptionsMenu(menu, inflater);
     }
 
+    @Override
+    public void onPrepareOptionsMenu(Menu menu) {
+        menu.getItem(0).setEnabled(!mInviteOperationInProgress); // here pass the index of save menu item
```
Comment (hypest) [inline WordPress/src/main/java/org/wordpress/android/ui/people/PeopleInviteFragment.java]: Minor nit: I guess the comment should refer to "Send", not "save".

File: WordPress/src/main/java/org/wordpress/android/ui/people/PeopleInviteFragment.java [hunk 4]
```diff
@@ -442,6 +454,10 @@ private void checkAndSend() {
             return;
         }
 
+        //set the  "SEND" option disabled
+        mInviteOperationInProgress = true;
+        getActivity().invalidateOptionsMenu();
```
Comment (hypest) [inline WordPress/src/main/java/org/wordpress/android/ui/people/PeopleInviteFragment.java]: This couple of lines (setting the boolean and then calling `invalidateOptionsMenu()`) seem to appear a few times so, may I suggest refactoring those into a function? Thanks!

**Comment by hypest** [comment]
Thanks for this PR @mzorz !

I left a few comments, most important of which are the rotation issue and the (non-existent) error message when bailing the `checkAndSend()` if offline.

I'm afraid that the latter of those means that the current implementation in this PR doesn't actually fix the issue in question (#4332) which is exactly about nothing meaningful happening when hitting SEND while offline. Am I missing something?

File: WordPress/src/main/java/org/wordpress/android/ui/people/PeopleInviteFragment.java (hunk 2 above)
Comment (mzorz) [inline WordPress/src/main/java/org/wordpress/android/ui/people/PeopleInviteFragment.java]: The implementation of `NetworkUtils.checkConnection()` actually shows a Toast when there is no network 👍

Comment (hypest) [inline WordPress/src/main/java/org/wordpress/android/ui/people/PeopleInviteFragment.java]: Aha, that's true! Forgot about that toast. Sorry for the unneeded proposal.

BTW, notice that when no connection, `checkAndSend()` doesn't get called, and prolly this is part of the root cause for #4332

**Comment by mzorz** [comment]
Thanks for pointing all these things out @hypest ! 💯 

> Aha, that's true! Forgot about that toast. Sorry for the unneeded proposal.
> BTW, notice that when no connection (and the rest of the steps in the parent issue), checkAndSend() doesn't get called, and prolly this is part of the root cause for #4332

So just to clarify: this PR aims at addressing both the exact case the issue describes in its title (being totally offline and under _that_ situation attempting to send an invitation, [which is addressed here](https://github.com/wordpress-mobile/WordPress-Android/pull/4335/files#diff-fdb8073841701f644a4de38798d117c9R438), just as you correctly pointed out in your last comment) but also the broader and tougher case when the connection is slow or intermittent (i.e. disabling the SEND button while an operation is in progress). In this PR the user is notified / given a visual cue of what is going on in both cases. :+1:

There is something a bit more interesting though: 

- while this commit https://github.com/wordpress-mobile/WordPress-Android/pull/4335/commits/6aab1bcb51e93236335b6eabf6c8cefaa138922c  addresses the particular issue of saving the flag when the fragment is destroyed/recreated, I’ve discovered we have a somewhat bigger issue here when the screen is rotated and we have  a background operation going on.
The problem manifests itself specifically [in this line](https://github.com/wordpress-mobile/WordPress-Android/blob/issue/4332-send-button-invite-offline/WordPress/src/main/java/org/wordpress/android/ui/people/PeopleInviteFragment.java#L486):

``` @Override
                    public void onSent(List<String> succeededUsernames, Map<String, String> failedUsernameErrors) {
                        if (!isAdded()) {
                            return;
                        }
``` 

There, if the fragment is not added to the Activity then we just return, without ever processing the results. It might be the case (and I managed to reproduce a couple of times) that a configuration change is taking place and so yet the fragment is not added, thus the processing never happens (but it should, as we need to represent the UI state after the operation has taken place to inform the user about how it all went). In this case, the `isAdded` call is just assuming we’re safe to not do anything if the fragment hasn’t been attached yet, which is of course wrong.
There are several ways to get away from this situation, the easiest of them in terms of code cleanliness being IMO retaining this fragment. Having checked we are already using `setRetainInstance(true)` elsewhere in our codebase I’ve done so in commit https://github.com/wordpress-mobile/WordPress-Android/pull/4335/commits/89f29552d5ad2cdd0ad4a1f088e75d68550fe58a.


- Another issue I’ve found while doing some more thorough testing on this is that sometimes when you hit “send” nothing happens (actually something does happen, but it gives you no clue about what is going on); and if you let it stay and hit SEND again, even an NPE can occur on slow network conditions (because of a funny state is produced). I followed the code a bit to find the `validateUsernames()` call actually makes a request to the WPCOM API, which might as well be a lengthy operation, so I placed some new calls to disable/enable the SEND button and avoid people repeatedly hitting it.
(see commit https://github.com/wordpress-mobile/WordPress-Android/pull/4335/commits/bb14fa05d8ce3de2dab412ddd92db9d3e816ad5c)

I think this PR is meaningful when you test on slow network conditions, this is something we should have in mind for any feature, no matter how apparently small it looks in relation to other parts of the app :)

Please would you take another look? thanks again for taking the time to have a closer look, and thanks in advance for giving a second pass on this PR @hypest ! 👍

**Comment by hypest** [comment]
> being totally offline and under that situation attempting to send an invitation, _which is addressed here_

Hmm, I'm afraid I'm not seeing the original issue being fixed. Actually, it still behaves as described in #4332 . Perhaps I'm doing something wrong. @mzorz , can you please confirm that #4332 is not happening on your side? Thanks!

File: WordPress/src/main/java/org/wordpress/android/ui/people/PeopleInviteFragment.java [hunk 5]
```diff
@@ -93,14 +102,21 @@ public void onSaveInstanceState(Bundle outState) {
         outState.putSerializable(KEY_USERNAME_RESULTS, mUsernameResults);
         outState.putString(KEY_ROLE, mRole);
         outState.putString(KEY_CUSTOM_MESSAGE, mCustomMessage);
+        outState.putBoolean(KEY_INVITE_IN_PROGRESS, mInviteOperationInProgress);
```
Comment (hypest) [inline WordPress/src/main/java/org/wordpress/android/ui/people/PeopleInviteFragment.java]: With `setRetainInstance(true)`, saving the `mInviteOperationInProgress` boolean is actually not needed as the Fragment instance is kept. Basically, same holds true for all these instance variables anyway. May I suggest removing this manual saving? Thanks!

**Comment by hypest** [comment]
Regarding the `setRetainInstance()` I won't lie to you, I'm not a fan and I usually think it's a shortcut we shouldn't take. In this case, since we're already devoting the extra time to fine-tune the UX on the SEND button behaviour, I think we should probably walk the extra mile and avoid `setRetainInstance`, for example using a Service instead. WDYT?

BTW, the original implementation of the feature (in which the UX can end up be poor on slower connections, that's true) didn't go down that "rabbit hole" since the user invites is not a top feature anyway. Also, the relevant backend endpoints have a "fire-and-forget" nature anyway so, "ignoring" the responses and redoing the calls is not a big problem.

**Comment by mzorz** [comment]
As discussed on Slack:

1) focusing on the issue #4332 only (fix here https://github.com/wordpress-mobile/WordPress-Android/pull/4335/commits/3af2f6ba391e6266ab4359dfcf490b5d6678b64b) and ditching setRetainInstance  (https://github.com/wordpress-mobile/WordPress-Android/pull/4335/commits/d89eacffd8f848e973876582b13c3afa6b5f4359)

2) giving this the ability to perform better on orientation changes will be left to another further PR

Thanks for your time @hypest ! Awesome chat ❤️

File: WordPress/src/main/java/org/wordpress/android/ui/people/PeopleInviteFragment.java [hunk 6]
```diff
@@ -317,6 +317,12 @@ private void validateAndStyleUsername(Collection<String> usernames, final Valida
         }
 
         if (usernamesToCheck.size() > 0) {
+
+            if (!NetworkUtils.checkConnection(getActivity())) {
```
Comment (hypest) [inline WordPress/src/main/java/org/wordpress/android/ui/people/PeopleInviteFragment.java]: The network check here seems to have the side effect of trying it even on device rotation, every time popping up a being-offline Toast even though the user didn't try to actually do anything.

Instead of having the check here, may I suggest moving the network check inside the `public void send()` method, [right before the SEND button disabling](https://github.com/wordpress-mobile/WordPress-Android/pull/4335/files#diff-fdb8073841701f644a4de38798d117c9R426)? I might be wrong with the suggestion so, please give it a shot or perhaps come up with a more correct solution. Thanks!

File: WordPress/src/main/java/org/wordpress/android/ui/people/PeopleInviteFragment.java [hunk 7]
```diff
@@ -108,13 +108,6 @@ public void onSaveInstanceState(Bundle outState) {
     }
 
     @Override
-    public void onCreate(Bundle savedInstanceState) {
-        super.onCreate(savedInstanceState);
-        // retain this fragment across configuration changes
-        setRetainInstance(true);
```
Comment (hypest) [inline WordPress/src/main/java/org/wordpress/android/ui/people/PeopleInviteFragment.java]: Removing the `setRetainInstance` has the side effect of the SEND button never getting re-enabled when the network calls finish. I guess the SEND button enabling/disabling should be removed as well, BUT: that would actually "cannibalise" this PR.

At this point, and basically aligning with the Slack chat we had, I think this PR should be left with `setRetainInstance` in and be allowed to continue the work for a better UX here and also fork the #4332 fix into a separate PR. WDYT?

**Comment by hypest** [comment]
Hey @mzorz , I enjoyed our chat as well! 👍 

I left a couple of new comments. Please, also have a look at the one that GitHub marks as "on an outdated diff" https://github.com/wordpress-mobile/WordPress-Android/pull/4335#discussion_r69865747, since it's actually a new comment. Thanks!

File: WordPress/src/main/java/org/wordpress/android/ui/people/PeopleInviteFragment.java (hunk 7 above)
Comment (mzorz) [inline WordPress/src/main/java/org/wordpress/android/ui/people/PeopleInviteFragment.java]: > Removing the setRetainInstance has the side effect of the SEND button never getting re-enabled when the network calls finish. I guess the SEND button enabling/disabling should be removed as well, BUT: that would actually "cannibalise" this PR.

😱 you're totally right!

File: WordPress/src/main/java/org/wordpress/android/ui/people/PeopleInviteFragment.java (hunk 6 above)
Comment (mzorz) [inline WordPress/src/main/java/org/wordpress/android/ui/people/PeopleInviteFragment.java]: sure thing! done here https://github.com/wordpress-mobile/WordPress-Android/pull/4335/commits/157eca164ab2029dcf80487af4ad9a9cd4700148 - your suggestion seems to work well 👍

**Comment by mzorz** [comment]
> At this point, and basically aligning with the Slack chat we had, I think this PR should be left with setRetainInstance in and be allowed to continue the work for a better UX here and also fork the #4332 fix into a separate PR. WDYT?

I 100% agree the easiest way out is using `setRetainInstance` 😄 
Also I have been reading quite a lot (*), I'm 100% confident it is harmless in this situation, also it is the right thing to do ("right" as in hitting the right balance, as compared to any other solution which so far seemed overkill for the case at hand). Anyway, I'm thinking of preparing another PR to handle this kind of situations in a better way using a headless fragment to handle network connections, that might be proper handling when there's more state to manage than just a couple of `EditText` components and an enabled/disabled button :)

(*) 
This is what the [official docs say about this](https://developer.android.com/guide/topics/resources/runtime-changes.html#RetainingAnObject):

It focuses on saving/loading back big chunks of data, but it also covers the situation of a network connection, which is our case:

```
“If restarting your activity requires […] re-establish a network connection […] then a full restart due to a configuration change might be a slow user experience”.
“In such a situation, you can alleviate the burden of reinitializing your activity by retaining a Fragment when your activity is restarted due to a configuration change”.
```

Here’s a [tutorial at the official developer.android.com page](https://developer.android.com/training/displaying-bitmaps/cache-bitmap.html#config-changes) where they encourage to use a retained fragment for the case of keeping large objects (caching bitmaps).

So basically, it's just another tool and like any tool I agree it would be foolish to use it indiscriminately but it's OK if correctly applied.

I might as well add a comment to the `setRetainedInstance(true)` call for passers-by to know it should be used with care and not fall into the rather lazy _"they used it there, I'll use it here too"_ situation.

**Comment by hypest** [comment]
> I 100% agree the easiest way out is using setRetainInstance

Please note that I wasn't suggesting that the `setRetainInstance` usage is the final solution I'd like to see. I was only highlighting the fact that just removing `setRetainInstance` actually _breaks_ the SEND enable/disable behaviour so, in the current incarnation of the PR it needs to be reinstated. But, I still think we need to explore better solutions than the current.

> in a better way using a headless fragment to handle network connections
and
> Here’s a tutorial at the official developer.android.com page where they encourage to use a retained fragment for the case of keeping large objects (caching bitmaps).

Exactly, "headless" is the key word here. The tutorial is about a headless fragment anyway. I don't mind retaining a fragment that is basically a cache mechanism. That's not the case with the fragment in the current implementation of this PR though which is quite heavy on UI operations as well.

I think all arguments that could be made have been made and there is no need for blocking the #4332 any longer. A better solution could be explored in a future PR. I've confirmed that the issue mentioned in #4332 is now fixed and so: :shipit:

**Comment by mzorz** [comment]
> we need to explore better solutions than the current.

I agree!

We still have this problem:

> Removing the setRetainInstance has the side effect of the SEND button never getting re-enabled when the network calls finish. 

I was in the middle of re-adding this when your comment and merge came in - Just pushed it to the (now new) branch `issue/4332-send-button-invite-offline`.

Want to merge that?

**Comment by hypest** [comment]
Hey @mzorz sure, can you please open a new PR with the fix? Thanks!

**Comment by mzorz** [comment]
Thanks, check #4362 please @hypest 
🙇
